src/Compiler/Service/TransparentCompiler.fs

```diff
@@ -0,0 +1,2105 @@
+            loadClosureOpt: LoadClosure option,
+            basicDependencies,
+            importsInvalidatedByTypeProvider: Event<unit>
+        ) =
+
+        node {
+            let diagnosticsLogger =
+                CompilationDiagnosticLogger("CombineImportedAssembliesTask", tcConfig.diagnosticsOptions)
+
+            use _ = new CompilationGlobalsScope(diagnosticsLogger, BuildPhase.Parameter)
+
+            let! tcImports =
+                node {
+                    try
+                        let! tcImports =
+                            TcImports.BuildNonFrameworkTcImports(
+                                tcConfigP,
+                                frameworkTcImports,
+                                nonFrameworkResolutions,
+                                unresolvedReferences,
+                                dependencyProvider
+                            )
+#if !NO_TYPEPROVIDERS
+                        // TODO: review and handle the event
+                        tcImports.GetCcusExcludingBase()
+                        |> Seq.iter (fun ccu ->
+                            // When a CCU reports an invalidation, merge them together and just report a
+                            // general "imports invalidated". This triggers a rebuild.
+                            //
+                            // We are explicit about what the handler closure captures to help reason about the
+                            // lifetime of captured objects, especially in case the type provider instance gets leaked
+                            // or keeps itself alive mistakenly, e.g. via some global state in the type provider instance.
+                            //
+                            // The handler only captures
+                            //    1. a weak reference to the importsInvalidated event.
+                            //
+                            // The IncrementalBuilder holds the strong reference the importsInvalidated event.
+                            //
+                            // In the invalidation handler we use a weak reference to allow the IncrementalBuilder to
+                            // be collected if, for some reason, a TP instance is not disposed or not GC'd.
+                            let capturedImportsInvalidated = WeakReference<_>(importsInvalidatedByTypeProvider)
+
+                            ccu.Deref.InvalidateEvent.Add(fun _ ->
+                                match capturedImportsInvalidated.TryGetTarget() with
+                                | true, tg -> tg.Trigger()
+                                | _ -> ()))
+#endif
+#if NO_TYPEPROVIDERS
+                        ignore importsInvalidatedByTypeProvider
+#endif
+                        return tcImports
+                    with
+                    | :? OperationCanceledException ->
+                        // if it's been canceled then it shouldn't be needed anymore
+                        return frameworkTcImports
```

Comment:
@0101 Are you sure it's safe to return this value here instead of rethrowing the exception? Can this value get cached somewhere?